### ISSUE TITLE
(tracking) Tree Data Filtering

### ISSUE THREAD

**Issue by ghiscoding** [ghiscoding/SlickGrid#1]
TODO
1. Seems that `filterMyFiles()` returns duplicate results, so for now I call `uniqueArray()` to remove any duplicates
   - would be nice to avoid having to dedupe but I can also live with it, not a huge problem
2. I manage to get the demo working, the `myFilter()` calls the `var tmpResult = filterMyFiles()` and then I do a `tmpResult.find()` to see if the object Id is in the array, I return true/false with that in mind
   - perhaps `filterMyFiles()` could simply store the object Ids? That might improve search quite a lot
3. The code requires `eval` of `Function('return '...')` which is unsafe, can that be replaced?

**Comment by Alex-ley** [comment]
See Pull Request #2 to remove the need to call ``uniqueArray(filteredChildrenAndParents);``

The next pull request will attempt to save instead an array of IDs to then be consumed by the ``myFilter()`` function.

**Comment by ghiscoding** [comment]
This is seriously awesome work, a real big thanks for all of your time spent on this. 

`const comparison = re.test(columnFilters[key]) && Function('return ' + a[key] + columnFilters[key])();`

I didn't know about the `Function`, it felt like an `eval` but I think it's because I misread the `eslint` warning. However could you explain me what it does? I'll do a log to see what it returns but I thought you might be able to explain it better. 

Does the `Function` just execute `a[key] + columnFilters[key]` and returns it? If so why not just write that code directly?

> Note: so far we accept only JavaScript conditionals and comparators like "<=3" or "!=3" but I noticed elsewhere that the library accepts "<>3" - I didn't address this yet but we could easily address it by either building an object of each conditional and comparator which then calls methods or we can just replace "<>" with "!=".

Yes my lib is really flexible on that side, I've added a bunch of these conditions and the goal for me is to drop your code into the lib and use what is already in place (if possible). I'm not too worried about that part and I also added what I named Compound Input Filters a year after, it combines a dropdown to choose which Operator and an input value, this way is even easier to implement and more UX friendly. Though I also added the 2 dots (..) to deal with ranges, like `5..25` does a range between 5 and 25. 

My lib is very flexible, I spent the past 2 years improve it and we use it for a few of our company projects. We can't live without data grids. Hope you'll be a user in the future ;)

![image](https://user-images.githubusercontent.com/643976/78900833-70155600-7a45-11ea-8602-73ae14e199c6.png)

**Comment by Alex-ley** [comment]
Thanks :)

First, we check if the filter is a "simple" text filter like "Map" with:
``const strContains = String(a[key]).toLowerCase().includes(columnFilters[key].toLowerCase());``

Then we check if the filter is a complex comparator filter like "<3" with:
``const re = /(?:(?:^|[-+<>=_*/])(?:\s*-?\d+(\.\d+)?(?:[eE][+-<>=]?\d+)?\s*))+$/;``
and if it is then run the second part of this line:
``const comparison = re.test(columnFilters[key]) && Function('return ' + a[key] + columnFilters[key])();``

I think some kind of parsing from the comparator characters and value to javascript is needed to evaluate the filter for true or false on each item.

Essentially, ``Function('return ' + a[key] + columnFilters[key])();`` would evaluate ``item.size < 3`` if the filter is ```{size: "<3"}"

The other way we could do this is to set up an object right? like:

```js
const objCompareParser = {
    "<": function(a,b) {return a < b},
    ">": function(a,b) {return a > b},
    "=": function(a,b) {return a == b},
    "!=": function(a,b) {return a != b},
    ...
}
```

Then use this like:
``const comparison = re.test(columnFilters[key]) &&  objCompareParser["<"](a[key], columnFilters[key])``

Where we would have to parse "<" from ``{size: "<3"}``

Or maybe there is a much simpler way I am overlooking? How did you do it in the past / how does the library do it?

**Comment by ghiscoding** [comment]
Oh ok thanks for the explanation, I'm starting to understand :)

So in my lib, I tried to decouple as much as I could and so what I've done is a method call to execute the filter condition check like this (this code [line](https://github.com/ghiscoding/Angular-Slickgrid/blob/master/src/app/modules/angular-slickgrid/services/filter.service.ts#L355))
```ts
if (!FilterConditions.executeMappedCondition(conditionOptions)) {
   return false;
}
```
which finds the proper condition to execute depending on the field type, operator, ...
for example the number filter condition (code [line](https://github.com/ghiscoding/Angular-Slickgrid/blob/master/src/app/modules/angular-slickgrid/filter-conditions/numberFilterCondition.ts))
```ts
export const numberFilterCondition: FilterCondition = (options: FilterConditionOption) => {
  const cellValue = parseFloat(options.cellValue);
  const searchTerms = Array.isArray(options.searchTerms) && options.searchTerms || [0];
  return testFilterCondition(options.operator || '==', cellValue, searchValue1);
};
```
which itself calls the `testFilterCondition` (code [line](https://github.com/ghiscoding/Angular-Slickgrid/blob/30cf1b4c9c0ed0d38f93acf27a15ad3d196a776b/src/app/modules/angular-slickgrid/filter-conditions/filterUtilities.ts#L30))
```ts
export const testFilterCondition = (operator: OperatorString, value1: any, value2: any): boolean => {
  switch (operator) {
    case '<':
    case 'LT': return (value1 < value2);
    case '<=':
    case 'LE': return (value1 <= value2);
 ...
```

... so all in all, I think I can safely replace your `.test()` and `Function` by my current implementation and that should work. I think...

**Comment by ghiscoding** [comment]
Because it's decoupled, it's also much easier to troubleshoot and fix filter condition issues... "got a problem with how string are compared?" Ok so just see how it does the job in [StringFilterCondition](https://github.com/ghiscoding/Angular-Slickgrid/blob/30cf1b4c9c0ed0d38f93acf27a15ad3d196a776b/src/app/modules/angular-slickgrid/filter-conditions/stringFilterCondition.ts).. want to add a new filter type check? Just add a new filter condition... etc... :)

But I obviously couldn't put that in the SO question because it was already way too long hehe. 
Nice to see some people are interested in the subject 😉 
Perhaps someone might find I've done a horrible job at decoupling it lol

**Comment by Alex-ley** [comment]
Yeah makes sense, no need to re-invent the wheel when you already have the extensive `testFilterCondition()`` function and the other caller functions etc. :)

I assume we could just call the following function as you say

```js
if (!FilterConditions.executeMappedCondition(conditionOptions)) {
   return false;
}
``` 

Probably from around line 120?: https://github.com/ghiscoding/SlickGrid/blob/c6db7a47e766bfe2118876cf36b90d6675834bc4/examples/example-tree-data.html#L122

**Comment by ghiscoding** [comment]
@Alex-ley 
I have a question, this might improve performance or I might not understanding it well. Does the creation of the `myObj` [here](https://github.com/ghiscoding/SlickGrid/blob/c6db7a47e766bfe2118876cf36b90d6675834bc4/examples/example-tree-data.html#L110) need to be done every time we do a search or is that a 1 time thing on the full dataset? If it's a 1 time thing, we can move that outside of the `filterMyFiles` and call it once after the dataset is sent to the DataView. Thought?

**Comment by Alex-ley** [comment]
This actually caused some unexpected behaviour when I switched from my sandbox in stack overflow to the real implementation in my local clone of the repo.

In SO, I was only thinking about a filter being called once essentially.

In the example-tree-data.html file this function is fired on every key up event so therefore we fire the event 3 times after another when we type "Map".

First, we search with ``{file: "M"}``, then ``{file: "Ma"}``, then finally ``{file: "Map"}``

This means that the second time we call it with ``{file: "Ma"}`` the data was already filtered by the ``{file: "M"}`` so only limited rows in the table are showing.

This means we create myObj again from a much smaller row number and so then subsequent searches are very quick.

I did consider to move it outside but I wasn't sure how everything was fitting together etc.

I think we would do it here, right after we prepare the data we could run:
https://github.com/ghiscoding/SlickGrid/blob/c6db7a47e766bfe2118876cf36b90d6675834bc4/examples/example-tree-data.html#L212-L214

```js
for (let i = 0; i < inputArray.length; i++) {
        myObj[inputArray[i].id] = inputArray[i];
        //as the filtered data is then used again as each subsequent letter
        //we need to delete the ._used property, otherwise the logic below
        //in the while loop (which checks for parents) doesn't work:
        //CHANGE DONT RUN NEXT LINE HERE BUT RATHER CLEAR ALL USED FLAGS EACH SEARCH/FILTER
       //delete myObj[inputArray[i].id]._used;
      }
```

But as you can see, we also need to clear all of the ._used flags (props) each search otherwise it won't work as expected.

another option is we make a copy of the myObj each search like ``const myObjCopy = {...myObj}`` but that's not that performant I don't think.

**Comment by ghiscoding** [comment]
Yes it does, again thanks a lot for all your explanation, I did try to move it after the data is created and once I ran it, it was showing weird behavior, it was showing only the map files without their parents (documents/pdf). 

So, I think I'll stick what what we have now then.
I must say, I really enjoy collaborating with you :)

**Comment by Alex-ley** [comment]
The final option would be forget the ._used prop flag and just accept that sometimes it will have duplicate parent folder entires in the final array.

Then simply run your dedupe function on that result. I assume that is also at least one loop though? So probably no better in performance.

Yeah it’s been fun working together on this. I learnt about a new library (SlickGrid) I’d never heard of (as well as your 2 wrapper/framework integration libraries) which look really cool.

I’m a bit of a data guy so I’ve often used HTML tables to display data and so might have some use cases for SlickGrid in the future. I’ll definitely keep it in mind.

I had been thinking of contributing to an open source library for a while so maybe I’ll take a deeper look into the library and think about contributing more. Would be great to collaborate further. I never used TypeScript before and never worked on a larger project (I’m a product manager who codes for fun and to automate stuff make little tools etc but not as the main focus of my job) and never written any serious unit tests 🙀😂 so I could for sure learn a lot by contributing more.

**Comment by ghiscoding** [comment]
@Alex-ley 
I'm looking at implementing this in my own lib and while adding some logs I found out that the duplicate values is still happening. 

I tried changing this line
```js
parent.__used || filteredChildrenAndParents.splice(len, 0, parent.id);
```
to this 
```js
if (!parent.__used) {
  filteredChildrenAndParents.push(parent.id);
}
```
that doesn't change anything though. BTW, isn't a push performing better than splice?
Mind taking another look? If not, I'm ok with a uniqueArray fn, I don't think it would bring down performance at all for an array of number.

**Comment by ghiscoding** [comment]
>  I never used TypeScript before

TypeScript is great since it adds typings, it avoids a lot of problems at coding time instead of at run time, it makes javascript feel much closer to other languages like C#. Very easy to learn, the basis of it is to add the type with colon after the variable (for example, if `var1` is a number, then `var1: number`, if it's an array it would be `var1: number[]` or `var1: Array<number>` and so on, you can also create interfaces). 

> I’m a bit of a data guy so I’ve often used HTML tables to display data and so might have some use cases for SlickGrid in the future. I’ll definitely keep it in mind.

If you're a data guy, you should definitely look into possibly using SlickGrid, it's definitely great for showing data. The project that I'm working on is to display Customer Quote, so a Tree Data View is great for that use case. There are so many things we do with this grid, it's just so customizable and fast that it works for every scenario. You've probably took a look at the multiple samples I've created for [Angular-Slickgrid](https://ghiscoding.github.io/Angular-Slickgrid/#/basic).

Another grid library which is even better than SlickGrid is [Ag-Grid](https://www.ag-grid.com/) but most of their features are paid only, and it's not cheap. We prefer to stay with Open Source (aka free), I spent a lot of time creating these lib, but it is still worth the time since we now use it in multiple projects with a few having customer facing.

**Comment by Alex-ley** [comment]
I wasn’t sure if order was important. So a splice basically means the results keep their order like:

```js
[
  Grandparent,
    Parent,
      Child,
  Grandparent,
    Parent,
      Child,
]
```

Push would mean the order would be:

```js
[
  Child
    Parent,
      Grandparent,
 Child,
    Parent,
      Grandparent,
]
```
So I used splice to insert in before each child.

But now i think of it couldn’t we just use unshift to add to the front of the array and we get the same as splice?

If you want to create an issue in your lib and assign or tag me I’ll take a look. I was logging out in the example-tree-data.html and I’m pretty certain there was no duplicates. So I don’t understand what’s going on now.

**Comment by ghiscoding** [comment]
The order is not important since the `myFilter` just check if the `id` is found in the array and then returns true/false. I've refactored and simplified the code in a few places, I don't think that I changed the functionality but who knows maybe I broke something. 

Sure I can create an issue and assign it to you, let me add you as collaborator as well, that will allow you to push directly if you want to.
